test/src/basic/graphobserver.cpp

```diff
@@ -0,0 +1,406 @@
+				}
+			}
+			AssertThat(GO.m_events.at(2 + 5 + 7).first, Equals(GOEvent::CLEAR));
+		});
+
+		it("observes Graph::operator=", []() {
+			Graph G1;
+			randomGraph(G1, 5, 7);
+			Graph G2;
+			LoggingGraphObserver GO(&G2);
+			G2 = G1;
+			AssertThat(GO.m_cleared, Equals(1));
+			AssertThat(GO.m_nodesAdded, Equals(5));
+			AssertThat(GO.m_edgesAdded, Equals(7));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+
+			AssertThat(GO.m_events.size(), Equals(1 + 5 + 7));
+			AssertThat(GO.m_events.at(0).first, Equals(GOEvent::CLEAR));
+			for (int i = 0; i < 5 + 7; i++) {
+				if (GO.m_events.at(i + 1).first != GOEvent::NODE_ADD) {
+					AssertThat(GO.m_events.at(i + 1).first, Equals(GOEvent::EDGE_ADD));
+				}
+			}
+		});
+
+		it("observes Graph::insert(Graph)", []() {
+			Graph G1;
+			randomGraph(G1, 5, 7);
+			Graph G2;
+			LoggingGraphObserver GO(&G2);
+			G2.insert(G1);
+			AssertThat(GO.m_cleared, Equals(0));
+			AssertThat(GO.m_nodesAdded, Equals(5));
+			AssertThat(GO.m_edgesAdded, Equals(7));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+
+			AssertThat(GO.m_events.size(), Equals(5 + 7));
+			for (int i = 0; i < 5 + 7; i++) {
+				if (GO.m_events.at(i).first != GOEvent::NODE_ADD) {
+					AssertThat(GO.m_events.at(i).first, Equals(GOEvent::EDGE_ADD));
+				}
+			}
+		});
+
+		it("observes Graph deconstruction", []() {
+			std::unique_ptr<Graph> G1 {new Graph()};
+			LoggingGraphObserver GO(G1.get());
+			AssertThat(GO.getGraph(), Equals(G1.get()));
+			AssertThat(GO.m_graph, Equals(G1.get()));
+			G1->newNode();
+			AssertThat(GO.m_nodesAdded, Equals(1));
+			G1.reset();
+			AssertThat(GO.getGraph(), Equals(nullptr));
+			AssertThat(GO.m_graph, Equals(nullptr));
+			AssertThat(GO.m_reregistered, Equals(1));
+			AssertThat(GO.m_cleared, Equals(0));
+			AssertThat(GO.m_nodesAdded, Equals(1));
+			AssertThat(GO.m_edgesAdded, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+
+			{
+				Graph G2;
+				GO.reregister(&G2);
+				AssertThat(GO.getGraph(), Equals(&G2));
+				AssertThat(GO.m_graph, Equals(&G2));
+				AssertThat(GO.m_reregistered, Equals(2));
+				G2.newNode();
+				AssertThat(GO.m_nodesAdded, Equals(2));
+			}
+			AssertThat(GO.getGraph(), Equals(nullptr));
+			AssertThat(GO.m_graph, Equals(nullptr));
+			AssertThat(GO.m_reregistered, Equals(3));
+			AssertThat(GO.m_cleared, Equals(0));
+			AssertThat(GO.m_nodesAdded, Equals(2));
+			AssertThat(GO.m_edgesAdded, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+		});
+
+		it("observes Graph::split and Graph::unsplit", []() {
+			Graph G1;
+			LoggingGraphObserver GO(&G1);
+			edge e = G1.newEdge(G1.newNode(), G1.newNode());
+			AssertThat(e->index(), Equals(0));
+			edge f = G1.split(e);
+			AssertThat(f->index(), Equals(1));
```

Comment:
And changing how indices work would probably also affect users' code that implicitly relies on their behaviour even if that is not explicitly documented. Users will consider any publicly observable behaviour as part of the API anyways, so it might be even nice to have these tests fail if we change something about that behaviour.